test/extended/apiserver/webhook.go

```diff
@@ -0,0 +1,333 @@
+package apiserver
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+
+	buildv1 "github.com/openshift/api/build/v1"
+	imagev1 "github.com/openshift/api/image/v1"
+	exutil "github.com/openshift/origin/test/extended/util"
+
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+	e2edeployment "k8s.io/kubernetes/test/e2e/framework/deployment"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	"k8s.io/utils/pointer"
+)
+
+var _ = g.Describe("[sig-api-machinery][Feature:AdmissionWebhook]", func() {
+	defer g.GinkgoRecover()
+
+	oc := exutil.NewCLI("apiserver")
+
+	g.It("should call a validating admission webhook when accessing OpenShift API Server", func() {
+		targetNamespace := oc.Namespace()
+		kubeClient := oc.AdminKubeClient()
+		servicePort := int32(8443)
+		containerPort := int32(8444)
+
+		annotateNamespace(kubeClient, targetNamespace, "webhook-marker")
+		createService(kubeClient, targetNamespace, "openshift-test-webhook", "openshift-test-webhook", servicePort, containerPort)
+		deployWebhookServer(kubeClient, imageutils.GetE2EImage(imageutils.Agnhost), targetNamespace, "openshift-test-webhook", "openshift-test-webhook", "openshift-test-webhook", containerPort)
+		removeWebhookConfiguration := createValidatingWebhook(kubeClient, targetNamespace, "webhook-marker", "webhook-marker", "openshift-test-webhook", "openshift-test-webhook", servicePort)
+		defer removeWebhookConfiguration()
+
+		openshiftAPIResources := []struct {
+			name     string
+			createFn func() (func(), error)
+		}{
+			{
+				name: "image stream",
+				createFn: func() (func(), error) {
+					is := &imagev1.ImageStream{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:   "openshift-test-webhook",
+							Labels: map[string]string{"webhook-marker": "true"},
+						},
+					}
+					_, err := oc.AdminImageClient().ImageV1().ImageStreams(targetNamespace).Create(context.TODO(), is, metav1.CreateOptions{})
+					removeImageStream := func() {
+						err = oc.AdminImageClient().ImageV1().ImageStreams(targetNamespace).Delete(context.TODO(), is.GetName(), metav1.DeleteOptions{})
+						e2e.ExpectNoError(err, "deleting an image stream %s in namespace %s", is.Name, targetNamespace)
+					}
+					return removeImageStream, err
+				},
+			},
+
+			{
+				name: "build config",
+				createFn: func() (func(), error) {
+					bc := &buildv1.BuildConfig{
```

Comment:
I need to check if there is a better way of finding and creating resources. I think I could try to reuse data provided by `OpenshiftEtcdStorageData` test.